File: codeforces/102503/D.cpp
#include "bits/stdc++.h"
#define DEBUG(X) { auto _X = (X); std::cerr << "L" << __LINE__ << ": " << #X << " = " << (_X) << std::endl; }
#define REP(i, a) for (int i = 0, _##i = (a); i < _##i; ++i)

int32_t main() {
    std::ios::sync_with_stdio(false);
    // read names.
    std::map< std::pair<std::string, std::string>, int > idOf;
    std::string a1, a2, b1, b2;
    int nId = 0;
    while (std::cin >> a1) {
        if (a1[0] == '-') break;

        std::cin >> a2 >> b1 >> b2;
        idOf[make_pair(a1, a2)] = nId;
        idOf[make_pair(b1, b2)] = nId;
        nId++;
    }

    std::string query;
    std::set<int> all;
    while (std::cin >> query && query[0] != 'E') {
        if (query[0] == '+') {
            std::string q1, q2;
            std::cin >> q1 >> q2;

            all.insert(idOf[make_pair(q1, q2)]);
        } else if (query[0] == '-') {
            std::string q1, q2;
            std::cin >> q1 >> q2;

            all.erase(idOf[make_pair(q1, q2)]);
        } else if (query[0] == 'F') {
            std::string q1, q2;
            std::cin >> q1 >> q2;

            int id = idOf[make_pair(q1, q2)];
            std::cout << (all.count(id) ? "FOUND" : "404") << '\n';
        } else if (query[0] == 'U') {
            std::cout << all.size() << '\n';
        }
    }
    return 0;
}

File: codeforces/102503/D.rs
use std::collections::{HashMap, HashSet};

#[derive(Default)]
struct Scanner {
	buffer: Vec<String>
}
impl Scanner {
	fn next<T: std::str::FromStr>(&mut self) -> T {
		loop {
			if let Some(token) = self.buffer.pop() {
				return token.parse().ok().expect("Failed parsing input")
			}
			let mut input = String::new();
			std::io::stdin().read_line(&mut input).expect("Failed reading input");
			self.buffer = input.split_whitespace().rev().map(String::from).collect();
		}
	}
}

fn main() {
	let mut scanner = Scanner::default();
	// read names.
	let mut id_of: HashMap<(String, String), i32> = HashMap::new();
	let mut n_id = 0;
	loop {
		let a1 = scanner.next::<String>();
		if a1.starts_with('-') {
			break;
		}

		let a2 = scanner.next::<String>();
		let b1 = scanner.next::<String>();
		let b2 = scanner.next::<String>();

		id_of.insert((a1, a2), n_id);
		id_of.insert((b1, b2), n_id);
		n_id += 1;
	}

	// Answer queries.
	let mut all: HashSet<i32> = HashSet::new();
	loop {
		let query = scanner.next::<String>();
		match query.as_str() {
			"END" => {
				break;
			}
			"+" => {
				let q1 = scanner.next::<String>();
				let q2 = scanner.next::<String>();
				if let Some(&id) = id_of.get(&(q1, q2)) {
					all.insert(id);
				}
			}
			"-" => {
				let q1 = scanner.next::<String>();
				let q2 = scanner.next::<String>();
				if let Some(id) = id_of.get(&(q1, q2)) {
					all.remove(id);
				}
			}
			"FIND" => {
				let q1 = scanner.next::<String>();
				let q2 = scanner.next::<String>();
				if let Some(id) = id_of.get(&(q1, q2)) {
					if all.contains(id) {
						println!("FOUND");
					} else {
						println!("404");
					}
				}
			}
			"UNION" => {
				println!("{}", all.len());
			}
			_ => {
				eprintln!("Unknown query {}", query);
			}
		}
	}
}
